### PR TITLE
Split packages into finer grained dependencies

### PR DESCRIPTION
**Pull request by hmaarrfk**
Well it just took me 30 mins to rebuild my scientific environment after downloading :/ on windows.
I think the big dependencies are: opencv, xarray, pyqt, dask
pyqt and opencv are probably the big offenders, but dask (pure python) is an easy target.

Comments are appreciated 

From https://github.com/conda-forge/dask-feedstock/issues/46 it seems like this is something that people generally want.

Checklist
* [x] Used a fork of the feedstock to propose changes
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/conda_smithy.html#how-to-re-render ) with the latest `conda-smithy`
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by hmaarrfk** [comment]
@conda-forge-admin, please rerender

**Comment by mrocklin** [comment]
cc @jakirkham 

I'd like to keep the name `dask` for what you call `dask-all`.  A naive user will type in `conda install dask` and expect everything to work (we get tons of issues about this for `pip`) while a developer will read the docs and find mention of `dask-core` (what you call `dask`)

**Comment by hmaarrfk** [comment]
Marthew. Dask points to dask all and bokeh.

I think that is just confusing. Ill remove the dask-all package.

**Comment by hmaarrfk** [comment]
@conda-forge-admin, please rerender

**Comment by hmaarrfk** [comment]
I swear this recipe used to work 3 weeks ago when I developed it :/ I'll have to look at it again when I have more time.

**Comment by hmaarrfk** [comment]
Seems to be something with conda-smithy on my computer. it really doesn't like the fact that there is more than 1 thing in the output section.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,45 +1,117 @@
...
+      host:
+        - python
+        - pip
+      run:
+        - python
+      run_constrained:
+        - cloudpickle >=0.2.1
+        - cytoolz >=0.7.3
+        - distributed >=1.23.1
+        - numpy >=1.11.0
```
Comment (hmaarrfk) [inline recipe/meta.yaml]: @mrocklin do you want to support people having `dask-core` and numpy<1.11 at the same time? Admitidly, I don't think conda-forge has numpy less than 1.11, but you answer would give guidance on how to handle other minimum version pinnings.

**Comment by mrocklin** [comment]
I'm probably not the person to answer these questions.  If @jakirkham has
time I recommend deferring to him.

**Comment by mrocklin** [comment]
My guess is that if people are using dask-core then there should be no
constraint on any other python library, other than the core CPython
constraints.

**Comment by jakirkham** [comment]
Have been wrestling with one point for a while. Namely should the `dask-core` feedstock remain separate or should it be merged into this one. Can see arguments both ways. Right now my leaning is towards keeping it separate. Otherwise we can potentially run into a cycle issue with `distributed` (as it requires `dask-core`).

When looking at this PR, I can see the value of using `run_constrained`. Though I can also imagine cases where this is problematic. For example, what if `distributed` needed a newer version of `toolz` or `dask-dataframe` needed a newer version of `numpy` than `dask-array`. In Dask's `setup.py`, these are easy to change in an orthogonal manner. It's not clear to me that `run_constrained` provides this level of flexibility, but I have not explored this detail. Certainly having each package specify its own constraints would address this.

**Comment by jakirkham** [comment]
Thanks for working on this, @hmaarrfk 😄

**Comment by hmaarrfk** [comment]
@jakirkham you provide good points for moving the dependencies away from run-constrained.

I mostly wanted to avoid the issue where people say: "well I have all the dependencies installed, why can't dask-array run" as they would be running on some old versions. Though, only somebody know understood dependencies would even try to do that, so I think i'm trying to anticipate an issue that would seldom happen.

I think `dask-core` should just be merged in here. By removing the `run_constrained` section, I'm pretty sure this ends up generating the same package as the other feedstock.

**Comment by hmaarrfk** [comment]
Where does `cytoolz` go?

**Comment by hmaarrfk** [comment]
Ah, the cycle issue you mean for potential rebuilds..... yeah... this wouldn't ever be able to be rebuilt as the test part of this recipe would require `distributed` to be run :/

**Comment by hmaarrfk** [comment]
this package is alot less fun without the ability to integrate building core. This would totally be useful, but it is definitely more effort to maintain.

I'm still for having the `run-constrained` version where you just constraint to the most common version. It reduces the number of lines you need to update on a version change.

**Comment by hmaarrfk** [comment]
Friendly ping @jakirkham and @mrocklin. Just checking if there is still general interest in this.

**Comment by mrocklin** [comment]
I'll admit that this is not a high priority for me.  Most users seem to be pretty well served by the existing system and it's pretty easy to maintain.  

That being said I'm not against it.  I'd be happy to leave the decision to @jakirkham

**Comment by mrocklin** [comment]
So what would the release process look like if we want to simultaneously release dask/dask and dask/distributed in lock-step?

**Comment by hmaarrfk** [comment]
hmm, you mean now that the bot automatically updates things, good question. I could remove the bot's auto update.

**Comment by hmaarrfk** [comment]
Note: reverted the bot auto updating

**Comment by mrocklin** [comment]
Sorry, I'll expand the question a bit.  Previously we had three feedstocks, `dask-core`, `distributed`, and `dask`.  When I wanted to update both dask and distributed at the same time I used to update dask-core, then update distributed and require that version, then update the dask metapackage.  Now that dask-core and dask are in the same feedstock I'm curious what this workflow would look like.

**Comment by hmaarrfk** [comment]
@mrocklin sorry, dask (metapackage) and dask-core are still different feedstocks. as per @jakirkham's previous comment about circular dependencies.

The update procedure is the same now. Unfortunately, splitting the packages without `run_constrained` means you now have to check versions in multiple places on a version update.

An earlier iteration had `run_constrained` in `dask-core` but was decided against.


P.S. I didn't know about the circular dependency before starting on this adventure.

**Comment by hmaarrfk** [comment]
I think im going to close this PR for the following reasons:

1. I’m mostly interested in dask-array
2. dask-array and delayed only have two additional dependencies, cytoolz and toolz in addition to numpy. These are jot too hard to add to my own packages dependencies. 
3. By keeping the dask-core package, this doesn’t simplify packaging anymore. This was the main thesis of this PR. 
4. The circular dependency on distributed makes the unified dask/dask-core package impossible. 

Thank you @mrocklin and @jakirkham for helping me out on this.

File: recipe/meta.yaml [hunk 2]
```diff
@@ -2,40 +2,85 @@
...
+      imports:
+        - dask.distributed
+        - distributed
+  - name: dask-delayed
+    build:
+      noarch: python
+    requirements:
+      run:
+        - dask-core  =={{ version }}
+        - toolz >=0.7.3
```
Comment (jakirkham) [inline recipe/meta.yaml]: Should we have `cytoolz` here as well? Asking as we have added it everywhere else and it's missing here.

**Comment by jakirkham** [comment]
@conda-forge-admin, please re-render.

**Comment by jakirkham** [comment]
Sorry for the long delay here, @hmaarrfk. One comment otherwise LGTM. Would be nice to resolve the merge conflicts here as well.

**Comment by jakirkham** [comment]
My understanding is 3 is the most important of these. Otherwise there is not much need. Though I may be misunderstanding the goal.

If we merged `dask-core`, `distributed`, and `dask` into one feedstock, then we might be able to get 1 (cycle absorbed within the feedstock). Not sure how well that would work, but I suppose that could be explored.

Not totally sure what you mean by 2. I guess we could put them in Jinja `set` statements in one spot. It's not obvious to me how much this helps. Maybe others can comment.

**Comment by hmaarrfk** [comment]
3 is  definitely important, but if the cost is really high to the dask
maintainers than it might not be worth it.

**Comment by hmaarrfk** [comment]
So `dask[array]` just became too hard to depend on for scikit-image. 
and `dask` too heavy of a dependency on conda-forge.

Just leaving it here as a point of reference.

https://github.com/conda-forge/scikit-image-feedstock/pull/52#issuecomment-541586526

**Comment by TomAugspurger** [comment]
What actually happened to make dask too hard of a dependency? Is that the opinion of all the scikit-image maintainers?

**Comment by hmaarrfk** [comment]
It was mostly my opinion as dask had been removed from a core dependency of the project.

If it were not for having to add a few dependencies with their minimum versions for `dask[array]` I would have added them.

1. We already have to contend with updating our minimum dependencies. Adding a few extra lines to manage dask's dependencies wasn't an endeavor we wanted to maintain.
2. Dask was only providing a minor improvement to the library.

That said, me typing out these comments makes me think I should just bite the bullet and update the dask dependency again in scikit-image again. (which I will do)

**Comment by mrocklin** [comment]
Just to be clear, `dask[array]` is just `dask-core`, `toolz`, and `numpy`.  Scikit-Image already depends on numpy, so this would amount to adding `dask-core` and `toolz` to your dependencies.

However, I agree that Dask isn't really necessary for scikit-image.  I think that it makes sense to drop it as a dependency.

**Comment by jakirkham** [comment]
FWIW we made some changes to how Dask is handling dependencies. So `dask[array]` is now just `dask` + `numpy`, which `scikit-image` and others using `dask` for array processing likely already require

https://github.com/dask/dask/blob/913a192fd3b35696883bdd4c4eeba2418b946b3a/setup.py#L8-L17

**Comment by hmaarrfk** [comment]
I guess we should update https://github.com/conda-forge/scikit-image-feedstock/blob/master/recipe/meta.yaml#L62 as well correct?

**Comment by jakirkham** [comment]
We didn't add `cytoolz` as a dependency (only `toolz`)

**Comment by jakirkham** [comment]
That said we could add this to `dask-core`. Let me see what people think

**Pull request closed without merging**